src/jit/optimizer.cpp

```diff
@@ -5711,34 +5711,76 @@ bool Compiler::optNarrowTree(GenTree* tree, var_types srct, var_types dstt, Valu
         switch (tree->gtOper)
         {
             case GT_AND:
+                noway_assert(genActualType(tree->gtType) == genActualType(op1->gtType));
                 noway_assert(genActualType(tree->gtType) == genActualType(op2->gtType));
 
-                // Is op2 a small constant than can be narrowed into dstt?
-                // if so the result of the GT_AND will also fit into 'dstt' and can be narrowed
-                if ((op2->gtOper == GT_CNS_INT) && optNarrowTree(op2, srct, dstt, NoVNPair, false))
+                GenTree* opToNarrow;
+                opToNarrow = nullptr;
+                GenTree** otherOpPtr;
+                otherOpPtr = nullptr;
+                bool canNotNarrowOperand;
+                canNotNarrowOperand = false;
+
+                // If 'dstt' is unsigned and one of the operands can be narrowed into 'dsst',
+                // the result of the GT_AND will also fit into 'dstt' and can be narrowed.
+                // The same is true if one of the operands is an int const and can be narrowed into 'dsst'.
+                if (!gtIsActiveCSE_Candidate(op2) && ((op2->gtOper == GT_CNS_INT) || varTypeIsUnsigned(dstt)))
                 {
-                    // We will change the type of the tree and narrow op2
+                    if (optNarrowTree(op2, srct, dstt, NoVNPair, false))
+                    {
+                        opToNarrow = op2;
+                        otherOpPtr = &tree->gtOp.gtOp1;
+                    }
+                    else
+                    {
+                        canNotNarrowOperand = true;
+                    }
+                }
+
+                if ((opToNarrow == nullptr) && !gtIsActiveCSE_Candidate(op1) &&
+                    ((op1->gtOper == GT_CNS_INT) || varTypeIsUnsigned(dstt)))
+                {
```

Comment:
It is possible that the above if statement set:
`canNotNarrowOperand = true;`
And then we still try to narrow  op1 here...
Perhaps this should be an `else if` block without the check for `(opToNarrow == nullptr)`

Reply:
That’s intentional. It’s enough to narrow just one operand to an unsigned type (or a constant operand) to be able to narrow GT_AND.

Reply:
The op2 case can fail, and set `canNotNarrowOperand = true;`
Then op1 can be successful and `set opToNarrow = op1;`
Then later ` if (canNotNarrowOperand)` will still return failure, due to canNotNarrowOperand being true

Reply:
If op1 case is successful, opToNarrow will be non-null and we will return true in the
```
if (opToNarrow != nullptr)
```
code.